### PR TITLE
raft: forget the leader when stepping down if it's safe

### PR DESCRIPTION
**Pull request by iskettaneh**
First commit is based on #136025

This PR makes the leader forget that it was the leader when stepping down if it's safe to do so. The places this PR changes:

1) When stepping down due to CheckQuorum.
2) When stepping down due to leadership transfer.
3) When stepping down due to becoming a learner.
4) After a restart.
5) After a VoteLoss.

Epic: none

Release note: None

### CODE REVIEW

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1274 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBXF1j32rk_ZFjLapuR:-OBXF1j32rk_ZFjLapuS:b-735il5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2fc71efd83d2c31cb1d0d58c7c756be57c1725b/pkg/raft/raft.go#L1274)):*
> ```Go
> 	r.tick = r.tickElection
> 	if lead == None {
> 		r.resetLead()
> ```

The only downside I see here is that defortify() now wouldn't advance electionElapsed because we are already defortified.

**Review by nvb — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, 3 of 3 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`pkg/raft/raft.go` line 1257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbTtbh7fs6BWtGCZOv:-OBbTtbh7fs6BWtGCZOw:b-xmlk3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1257)):*
> ```Go
> 		// Once this state is stabilized (within and above pkg/raft), we can remove
> 		// this special case.
> 		r.lead = None
> ```

We can remove this manual `r.lead` assignment now.
___
*[`pkg/raft/raft.go` line 1263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbUape7ketMNAbIxzr:-OBbUape7ketMNAbIxzs:b-rbhftf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1263)):*
> ```Go
> 	r.reset(term)
> 	r.tick = r.tickElection
> 	if lead == None {
> ```

This second commit is essentially a reversion of https://github.com/cockroachdb/cockroach/pull/129100/commits/b1053d431476abb3c26da52461fe1ba69c0f3d3e. Should we just undo that commit?
___
*[`pkg/raft/raft.go` line 2714 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbVaer51rwqnrXJbcZ:-OBbVaer51rwqnrXJbc_:b-k7gu9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b70f6eadb3d52d666ef5d0379b7645a103ad97a/pkg/raft/raft.go#L2714)):*
> ```Go
> 		// and propose that change. Therefore, it should be safe at this point to
> 		// forget that we were the leader at this term and step down.
> 		r.becomeFollower(r.Term, None)
> ```

Should we make the same change in the `VoteLost` case of `stepCandidate`?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbUape7ketMNAbIxzr:-OBbmRsU1vZZcPnoV8r-:b-cfw4ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1263)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This second commit is essentially a reversion of https://github.com/cockroachdb/cockroach/pull/129100/commits/b1053d431476abb3c26da52461fe1ba69c0f3d3e. Should we just undo that commit?
</blockquote></details>

The second commit kinda assumed that if you explicitly step down with `lead==None`, you want to DeFortify (setting LeadEpoch to 0). This helps avoid assertions where we know that it's safe to step down while forgetting the leader (for example due to becoming a learner, or due to CheckQuorum). In those cases, we might be in a situation where r.lead == 0 but r.leadEpoch != 0. 

Shall we explicitly call DeFortify() before stepping down in those cases?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbTtbh7fs6BWtGCZOv:-OBfJqWE0O16QB0Nllc3:b-eg2bb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1257)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can remove this manual `r.lead` assignment now.
</blockquote></details>

I left a TODO to call r.resetLead() inside r.reset() once we verified that it's safe to do so (all places that stepdown to the same term can defortify at that point). Once we do that, we canm remove this manual assignment
___
*[`pkg/raft/raft.go` line 1263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbUape7ketMNAbIxzr:-OBfK8Si9erYLq4YOA0V:b7fka5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1263)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The second commit kinda assumed that if you explicitly step down with `lead==None`, you want to DeFortify (setting LeadEpoch to 0). This helps avoid assertions where we know that it's safe to step down while forgetting the leader (for example due to becoming a learner, or due to CheckQuorum). In those cases, we might be in a situation where r.lead == 0 but r.leadEpoch != 0. 

Shall we explicitly call DeFortify() before stepping down in those cases?
</blockquote></details>

I left a TODO to call r.resetLead() inside r.reset() once we verified that it's safe to do so (all places that stepdown to the same term can defortify at that point). What do you think?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OBbTtbh7fs6BWtGCZOv:-OBfMys3818WL3v7yj2W:b-942vky) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9201d22585fbc90b86c14e129c20ee441f84c8a/pkg/raft/raft.go#L1257)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I left a TODO to call r.resetLead() inside r.reset() once we verified that it's safe to do so (all places that stepdown to the same term can defortify at that point). Once we do that, we canm remove this manual assignment
</blockquote></details>

Nvm, it's safe to remove it now since I am calling  resetLead() if lead==none, and we already know that leadEpoch==0 inside that if-statement. I removed it.

**Review by arulajmani — COMMENTED** [review]
Intermediate comments. Flushing them so you can address them while you make the changes we discussed synchronously. 

> Reviewed 4 of 4 files at r1, 8 of 8 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-:-OCFnYCMB6ZRwKL1jCFB:b-nolmkq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OC-WD2V4D8hglRUyqNi:-OC-WD2V4D8hglRUyqNj:be5mddl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e92686ad130255faf8f3f3f95e35ea2a49f8566d/pkg/raft/raft.go#L1267)):*
> ```Go
> 	r.logger.Infof("%x became follower at term %d", r.id, r.Term)
> 
> 	// If we should broadcast a DeFortify message, we will do it in the next
> ```

This comment doesn't seem great -- it's written in the context of the code that existed previously, not what the code does/will do after this patch merges. I would reword this as:

```
Start de-fortifying eagerly so we don't have to wait out a full heartbeat timeout before sending the first de-fortification message.
```

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OC-WD2V4D8hglRUyqNi:-OCK5nPN7RbQObthU0nJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e92686ad130255faf8f3f3f95e35ea2a49f8566d/pkg/raft/raft.go#L1267)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This comment doesn't seem great -- it's written in the context of the code that existed previously, not what the code does/will do after this patch merges. I would reword this as:

```
Start de-fortifying eagerly so we don't have to wait out a full heartbeat timeout before sending the first de-fortification message.
```
</blockquote></details>

Done.

**Review by arulajmani — COMMENTED** [review]
Mostly LGTM

> Reviewed 16 of 16 files at r21, 4 of 4 files at r22, 1 of 1 files at r23, 2 of 2 files at r24, 4 of 4 files at r25, 3 of 3 files at r26, 2 of 2 files at r27, 1 of 1 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-:-OCKoSV718VHKmXfp0tD:bgmtcxr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1276 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OCKhK2S7Hh1wrTtcPEZ:-OCKhK2S7Hh1wrTtcPE_:b9c2hwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f58438e044f082de5c9a8343f18ee91595671ad8/pkg/raft/raft.go#L1276)):*
> ```Go
> 	r.reset(term)
> 	r.setLead(lead)
> 	r.logger.Infof("%x became follower at term %d", r.id, r.Term)
> ```

nit: consider adding a `r.logger.Debugf("%x reset  election elapsed to %d", r.id, r.electionElapsed)` just so we're able to see that the election elapsed is reset back to 0 after the call to de-fortify in the data driven tests?
___
*[`pkg/raft/raft.go` line 491 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OCKo3a69803FN1x7oZ1:-OCKo3a69803FN1x7oZ2:b-luwrgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14c7305c6346389ce67824aa0cf760ef489692f/pkg/raft/raft.go#L491)):*
> ```Go
> 
> 	if r.lead == r.id {
> 		// If we were the leader, we must have waited out the leaderMaxSupported
> ```

nit: leadMaxSupported.

Seperately, could you add words for how we know that we've waited out the leadMaxSupported at this point? Also, worth adding words about why we forget the leader when becoming a follower (even though it is safe, it's unclear to a reader why we're doing so).
___
*[`pkg/raft/raft.go` line 498 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OCKnwmT-e9B0sxOe0dR:-OCKnwmT-e9B0sxOe0dS:b-gctii0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14c7305c6346389ce67824aa0cf760ef489692f/pkg/raft/raft.go#L498)):*
> ```Go
> 	} else {
> 		// If we weren't the leader, we should NOT forget who the leader is to avoid
> 		// regressing the leaderMaxSupported.
> ```

nit: leadMaxSupported. 

Separately, can you add words on how the leadMaxSupported could regress if we were to forget the leader?

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo comment.

> Reviewed 35 of 35 files at r29, 1 of 1 files at r30, 2 of 2 files at r31, 4 of 4 files at r32, 3 of 3 files at r33, 2 of 2 files at r34, 1 of 1 files at r35, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-:-OCL7zUL37ZQEgwfhB3p:b-913mt2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 501 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OCL56V0BgWCQUDdluLf:-OCL56V0BgWCQUDdluLg:bre2rny) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b5705624ce89f0763c308015571ea8a96a499a/pkg/raft/raft.go#L501)):*
> ```Go
> 	} else {
> 		// If we weren't the leader, we should NOT forget who the leader is to avoid
> 		// regressing the leadMaxSupported. We can't just forget the leader because
> ```

Something about this looks off. I'd say something like:

```
// If we weren't the leader, we should NOT forget who the leader is to avoid
// regressing the leadMaxSupported. We can't just forget the leader because
// we might have been fortifying a leader before the restart and need to 
// uphold our fortification promise after the restart as well. To do so, we need
// to remember the leader and the fortified epoch, otherwise we may vote for
// a different candidate or prematurely call an election, either of which could
// regress the LeadSupportUntil.
```

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 501 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/134990#-OCL56V0BgWCQUDdluLf:-OCYbpG3AN9ggIlKKfuI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b5705624ce89f0763c308015571ea8a96a499a/pkg/raft/raft.go#L501)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Something about this looks off. I'd say something like:

```
// If we weren't the leader, we should NOT forget who the leader is to avoid
// regressing the leadMaxSupported. We can't just forget the leader because
// we might have been fortifying a leader before the restart and need to 
// uphold our fortification promise after the restart as well. To do so, we need
// to remember the leader and the fortified epoch, otherwise we may vote for
// a different candidate or prematurely call an election, either of which could
// regress the LeadSupportUntil.
```
</blockquote></details>

Done.

**Comment by iskettaneh** [comment]
TFTR!

bors r+
